scene/resources/material.cpp

```diff
@@ -380,6 +418,112 @@ bool ShaderMaterial::_property_get_revert(const StringName &p_name, Variant &r_p
 	return false;
 }
 
+void ShaderMaterial::_validate_uniform_struct_members(const Dictionary &p_default, Dictionary &p_current) const {
+	// Removes invalid members.
+	{
+		List<Variant> current_keys;
+		p_current.get_key_list(&current_keys);
+		for (const Variant &key : current_keys) {
+			if (!p_default.has(key)) {
+				p_current.erase(key);
+			}
+		}
+	}
+
+	List<Variant> default_keys;
+	p_default.get_key_list(&default_keys);
+	for (const Variant &key : default_keys) {
+		if (!p_current.has(key) || p_current[key].get_type() != p_default[key].get_type()) {
+			p_current[key] = p_default[key];
+		} else {
+			switch (p_current[key].get_type()) {
+				case Variant::PACKED_FLOAT32_ARRAY: {
+					const PackedFloat32Array &default_array = (PackedFloat32Array)p_default[key];
+					const PackedFloat32Array &array = (PackedFloat32Array)p_current[key];
+
+					if (array.size() != default_array.size()) {
+						PackedFloat32Array new_arr;
+						new_arr.resize(default_array.size());
+
+						const float *r = array.ptr();
+						float *w = new_arr.ptrw();
+						for (int i = 0; i < default_array.size(); i++) {
+							if (i < array.size()) {
+								w[i] = r[i];
+							} else {
+								w[i] = 0.0f;
```

Comment:
These could be filled in instead with values from the default array since we're still inside of it.